File: 907-KokoEatingBananas/907-KokoEatingBananas.cpp
// Last updated: 04/04/2026, 23:32:50
class Solution {
public:
    bool help(vector<int>& piles,int x,int h) {
        long time = 0;
        for(int num : piles) time += (num + x - 1L)/x;
        return time<=h;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int start = 1;
        int end = *max_element(piles.begin(),piles.end());
        while(start <= end) {
            int mid = start + (end-start)/2;
            if(help(piles,mid,h)) {
                end = mid-1;
            } else {
                start = mid+1;
            }
        }
        return start;
    }
};

File: 907-KokoEatingBananas/907-KokoEatingBananas.java
// Last updated: 04/04/2026, 23:32:46
class Solution {
    private boolean canComplete(int[] piles,int banasPerHr,int h) {
        long time = 0;
        for(int bananas : piles) {
            time += (bananas+banasPerHr-1L)/banasPerHr;
        }
        return time <= h;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int start = 1;
        int end = Integer.MIN_VALUE;
        for(int bananas : piles) end = Math.max(end, bananas);
        while(start <= end) {
            int banasPerHr = start + (end-start)/2;
            if(canComplete(piles, banasPerHr, h)) {
                end = banasPerHr-1; // decrease speed 
            } else {
                start = banasPerHr + 1; // increase speed
            }
        }
        return start;
    }
}